### PR TITLE
chore: use build system to create the VMs on CI

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
This PR is replacing the custom Groovy code in the pipeline used to create the cloud machines. Instead, we are using the make goals we added in #2454

## Why is it important?
It will unify the local experience with the CI experience, as both will be using the very same commmands

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have run the Unit tests (`make unit-test`), and they are passing locally
- [ ] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## Related issues

- Continuation of #2454

### CODE REVIEW

**Comment by cachedout** [comment]
> It will unify the local experience with the CI experience, as both will be using the very same commmands

💯 for this technique. Bravo!

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -205,24 +207,13 @@ pipeline {
                 script {
                   def stackWorkspace = "${env.WORKSPACE}/${env.BASE_DIR}"
                   // Deploy the test infrastructure
-                  sh "ssh-keygen -b 4096 -t rsa -f ${stackWorkspace}/e2essh -q -N \"\" "
-                  def stackMachine = getMachineInfo('stack')
+                  sh "ssh-keygen -b 4096 -t rsa -f ${E2E_SSH_KEY} -q -N \"\" "
```
Comment (mdelapenya) [inline .ci/Jenkinsfile:210]: The key will be transferred to the nodes in the `stash` built-in step

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -334,13 +325,43 @@ def checkSkipTests() {
...
+      "ANSIBLE_HOST_KEY_CHECKING=False",
+    ]){
+      withVaultToken(){
+        withEnvMask(vars: [
+          [var: "AWS_ACCESS_KEY_ID", password: awsAuthObj.access_key],
+          [var: "AWS_SECRET_ACCESS_KEY", password: awsAuthObj.secret_key]
+        ]) {
+          retryWithSleep(retries: 3, seconds: 5, backoff: true){
+            withOtelEnv() {
+              sh("make -C .ci setup-env") // make sure the environment is created
```
Comment (mdelapenya) [inline .ci/Jenkinsfile:358]: The .venv will be transferred to the nodes in the `stash` built-in step. It increases the stash/unstash time, so we could improve that. 

OTOH, the nodes will share python env

**Comment by mdelapenya** [comment]
Ok, this is now working, but for some reason I see constant disconnections of AWS instances from the Jenkins workers:

>[2022-05-11T17:08:22.611Z]     "msg": "Failed to connect to the host via ssh: OpenSSH_8.2p1 Ubuntu-4ubuntu0.4, OpenSSL 1.1.1f  31 Mar 2020\r\ndebug1: Reading configuration data /var/lib/jenkins/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug3: /etc/ssh/ssh_config line 19: Including file /etc/ssh/ssh_config.d/50-cloudimg-settings.conf depth 0\r\ndebug1: Reading configuration data /etc/ssh/ssh_config.d/50-cloudimg-settings.conf\r\ndebug1: /etc/ssh/ssh_config line 21: Applying options for *\r\ndebug2: resolve_canonicalize: hostname 3.134.115.217 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug1: Control socket \"/var/lib/jenkins/workspace/2e-tests_e2e-testing-mbp_PR-2490/.ansible/cp/271d02cce5\" does not exist\r\ndebug2: ssh_connect_direct\r\ndebug1: Connecting to 3.134.115.217 [3.134.115.217] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 3.134.115.217 port 22: Connection timed out\r\nssh: connect to host 3.134.115.217 port 22: Connection timed out",

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -536,6 +502,11 @@ def generateFunctionalTestStep(Map args = [:]){
     goArch = "arm64"
   }
 
+  // sanitize tags to create the file
+  def sanitisedTags = tags.replaceAll("\\s","_")
+  sanitisedTags = sanitisedTags.replaceAll("~","")
+  sanitisedTags = sanitisedTags.replaceAll("@","")
+
   def githubCheckSha1 = params.GITHUB_CHECK_SHA1?.trim() ? params.GITHUB_CHECK_SHA1 : ''
   def githubCheckRepo = params.GITHUB_CHECK_REPO?.trim() ? params.GITHUB_CHECK_REPO : ''
```
Comment (v1v) [inline .ci/Jenkinsfile:511]: IIUC, those env variables are already defined in the top-level environment section with `trim()`:

```suggestion
  def githubCheckSha1 = env.GITHUB_CHECK_SHA1
  def githubCheckRepo = env.GITHUB_CHECK_REPO
```

Comment (mdelapenya) [reply .ci/Jenkinsfile:511]: Yes, I saw that. But I'm consistently seeing a `"null"` value being passed to Ansible: `GithubCommitSha=null`

>[2022-07-04T05:35:35.576Z] time="2022-07-04T05:35:34Z" level=info msg="Initial artifact versions defined" BeatVersion=7.17.5-b5794350-SNAPSHOT BeatVersionBase=7.17.5-b5794350-SNAPSHOT ElasticAgentVersion=7.17.5-b5794350-SNAPSHOT GithubCommitSha=null GithubRepository=elastic-agent KibanaVersion=7.17.5-b5794350-SNAPSHOT StackVersion=7.17.5-b5794350-SNAPSHOT

URL: https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp%2F7.17/detail/7.17/818/pipeline

**Comment by kuisathaverat** [comment]
>I'm frustrated demonstrating that consistent behavior that I'm not able to explain why. Do you see anything wrong there?

The post/cleanup is executed right after you launch the parallel tasks, it does not wait for the parallel stages that are in a new context outside that stage.

```
2e2 Stage -> Parallel stages -> rest of the pipeline
                  -> Post clean up    
```

**Comment by mdelapenya** [comment]
> The post/cleanup is executed right after you launch the parallel tasks, it does not wait for the parallel stages that are in a new context outside that stage.
> 
> ```
> 2e2 Stage -> Parallel stages -> rest of the pipeline
>            -> Post clean up    
> ```

I think we can move the destroy code to the post-stage of the pipeline, and destroy all the machines at once, not on every stage.

Reading your answer I wonder if the destruction of the machines happens because other parallel stages are destroying other stages' instances 🤔

**Comment by kuisathaverat** [comment]
>Reading your answer I wonder if the destruction of the machines happens because other parallel stages are destroying other stages' instances 🤔

no, how I have said that the stage `End-To-End Tests` ends before the parallel tasks are executed so it run the destroy meanwhile the parallel tasks are executed.  the stage `End-To-End Tests` do not wait for the parallel tasks to end.

**Comment by mdelapenya** [comment]
> * `ciBuild` function is almost everywhere, I wonder what's the reason for using per make execution rather than globally?

Maybe the name is misleading... could it be just `runBuildCommand` for clarity? It's main goal is to be a wrapper to run make commands with the right environment (python venv)

> * `ciBuild` calls `sh("make -C .ci setup-env")` , will that command change anything if it's executed a few times in a row for the same `test` within the same `node`?

I've thought about that a few times, as it's adding _only_ 6 seconds each time it's called. I was tempted to add a check to verify if the .venv dir is created to avoid calling it

**Comment by mdelapenya** [comment]
@kuisathaverat @v1v I think I resolved the destruction error, narrowing down the tag name to be used when creating/destroying the instance.

It's now working. The errors are related to the same errors that are currently happening on main, caused by the 30-day retention policy for the current pinned artifacts version :(

**Review by v1v — APPROVED** [review]

**Comment by mdelapenya** [comment]
@mergifyio refresh

**Pull request merged**